### PR TITLE
Extend `UnknownFlags` to `UnknownFlagsHandling` and introduce `PassUnknownFlagToArgs` (#337)

### LINKED ISSUE #337: Preserve unknown flags
`FlagSet.ParseErrorsWhitelist.UnknownFlags` ignores unknown flags, but throws away them.
I want `pflag` just skip them and pass them to `Args()`.

Input:
```
--known-flag known-flag-value --unknown-flag-value arg0 arg1 arg2
```

Current output via `Args()`:
```
arg1 arg2
```
arg0 is treated as value for unknown-flag-value

Wanted output via `Args()`:
```
--unknown-flag-value arg0 arg1 arg2
```

Background:
I want to create a utility wrapper for existing command. It accepts extra flags and change how to call the command.
It's used like this:
```
mygrep --my-extra-flag1 --my-extra-flag2 my-extra-flag2-value -i -e search-text file1 file2
```

### PR DESCRIPTION
**Pull request by ikedam**
Close #337

Alternative to #338.

`FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling=PassUnknownFlagToArgs` behaves like this:

* Input:
    ```
    -xayb -c -z --known-flag known-flag-value --unknown-flag arg0 arg1 arg2
    ```
    * -x, -y, -z and  `--unknown-flag` are unknown
* Args() results:
    ```
    -xy -z --unknown-flag arg0 arg1 arg2
    ```

This request deprecates `FlagSet.ParseErrorsAllowlist.UnknownFlags` and extends it to `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling`.

* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling = ErrorOnUnknownFlag` (Default)
    * Same to `FlagSet.ParseErrorsAllowlist.UnknownFlags = false`
    * Treats unknown flags error.
* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling = IgnoreUnknownFlag`
    * Same to `FlagSet.ParseErrorsAllowlist.UnknownFlags = true`
    * Ignore unknown flags.
* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling=PassUnknownFlagToArgs`:
    * Pass unknown flags to `Args()` (described above)

Backward compatibilities:

* `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling` is not set (`ErrorOnUnknownFlag`) and `FlagSet.ParseErrorsAllowlist.UnknownFlags = true`:
    * Same to `FlagSet.ParseErrorsAllowlist.UnknownFlagsHandling = ErrorOnUnknownFlag`
    * Ignore unknown flags.
* Also same for `ParseErrorsWhitelist`

### CODE REVIEW

File: flag_test.go [hunk 1]
```diff
@@ -523,6 +523,111 @@ func testParseWithUnknownFlags(f *FlagSet, t *testing.T) {
 	}
 }
 
+func testParseWithUnknownFlagsAndPassToArgs(f *FlagSet, t *testing.T) {
+	if f.Parsed() {
+		t.Fatal("f.Parse() = true before Parse")
+	}
+	f.ParseErrorsAllowlist.UnknownFlagsHandling = UnknownFlagsHandlingPassUnknownToArgs
```
Comment (ikedam) [inline flag_test.go:530]: This test code is same to https://github.com/spf13/pflag/pull/338/files#diff-7a7385a10188cdaa9ac18effc375607e65672f55084b8049b6d3f44a6ab70c51 except this line.

File: flag.go [hunk 2]
```diff
@@ -137,10 +137,42 @@ const (
...
+const (
+	// UnknownFlagsHandlingErrorOnUnknown will return an error if an unknown flag is found
+	UnknownFlagsHandlingErrorOnUnknown UnknownFlagsHandling = iota
+	// UnknownFlagsHandlingIgnoreUnknown will ignore unknown flags and continue parsing rest of the flags
+	UnknownFlagsHandlingIgnoreUnknown
+	// UnknownFlagsHandlingPassUnknownToArgs will treat unknown flags as non-flag arguments.
+	// Combined shorthand flags mixed with known ones and unknown ones results
+	// combined flags only with unknown ones.
+	// E.g. -fghi results -gh if only `f` and `i` are known.
+	UnknownFlagsHandlingPassUnknownToArgs
```
[the remark this answers is not in the document]
Comment (tomasaschan) [reply flag.go]: Maybe `ErrorOnUnknownFlag`? Along with `IgnoreUnknownFlag` and `PassUnknownFlagToArgs`. I don't think they can be made shorter than that without losing meaning, but I don't think we need to be overly cautious about naming conflicts here.

If you really want to guard against it, another variant is to namespace these somehow - either with a _struct value_ where the enum values are fields on the struct, or by just moving them to an appropriately named subpackage.

File: flag.go [hunk 3]
```diff
@@ -137,10 +137,42 @@ const (
...
+func (a *ParseErrorsAllowlist) getUnknownFlagsHandling() UnknownFlagsHandling {
+	// if UnknownFlagsHandling is set, use it
+	if a.UnknownFlagsHandling != UnknownFlagsHandlingErrorOnUnknown {
+		return a.UnknownFlagsHandling
+	}
+
+	if a.UnknownFlags {
+		return UnknownFlagsHandlingIgnoreUnknown
+	}
+	return UnknownFlagsHandlingErrorOnUnknown
```
Comment (tomasaschan) [inline flag.go]: Nice!

**Review by tomasaschan — COMMENTED** [review]
I like this version a lot better than #338.

(The passing-error-as-control-flow thing smells a little, but the `parseLongArg` method was a mess before this change, so that's not your fault... Something to look out for in 2.0, I guess.)

Needs a rebase (I haven't looked into why) and probably shouldn't be merged until we've decided we don't have anything else we want to get into 1.0 before cutting 1.1.

**Comment by ikedam** [comment]
First, rebased and resolved the conflict.
Conflicted with #446 .

Needs to support both `ParseErrorsAllowlist` and `ParseErrorsWhitelist`, and resolved as https://github.com/spf13/pflag/pull/440/changes/d1d7aeebabe5cb8f98a083bc82685272c6083c1b .
The conflict was left unresolved in the previous commit and fixed in this one so reviewers can see exactly how it was resolved.

File: flag.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ikedam) [reply flag.go]: Renamed in https://github.com/spf13/pflag/pull/440/changes/c0bdd631c7115eaabfce7d6c38f09cb42da8909e

File: flag.go [hunk 4]
```diff
@@ -999,22 +1058,25 @@ func (f *FlagSet) parseLongArg(s string, args []string, fn parseFunc) (a []strin
...
+		case unknownFlagsHandling == IgnoreUnknownFlag:
 			// --unknown=unknownval arg ...
 			// we do not want to lose arg in this case
 			if len(split) >= 2 {
 				return a, nil
 			}
 
 			return stripUnknownFlagValue(a), nil
+		case unknownFlagsHandling == PassUnknownFlagToArgs:
+			return a, &unknownFlagError{
```
Comment (tomasaschan) [inline flag.go:1077]: I don't understand why the previous case calls `stripUnknownFlagValue` and this is one does not.

File: flag.go [hunk 5]
```diff
@@ -1077,6 +1139,20 @@ func (f *FlagSet) parseSingleShortArg(shorthands string, args []string, fn parse
...
+				err = &unknownFlagError{
+					UnknownFlags: shorthands,
+				}
+				return
+			}
+			// '-fgh': pass only the first switch
+			err = &unknownFlagError{
+				UnknownFlags: shorthands[0:1],
+			}
+			return
```
Comment (tomasaschan) [inline flag.go:1155]: What about `-fg=g` were both f and g are shorthands, but only g accept a value? Or `-fgh` (same)?

**Review by tomasaschan — APPROVED** [review]
I'm sorry this took so long to get to reviewing. I have a couple of questions, but given this behavior is opt-in I think it would be OK to merge this to make it easier to test in real applications, and roll forward when more edge cases pop up.

Let me know if you prefer a merge now or to address those questions first.

File: flag.go (hunk 4 above)
Comment (ikedam) [reply flag.go:1077]: We can consider the case of `--unknown somevalue`.
`--unknown` is an unknown flag, and we cannot reliably tell whether `somevalue` is a flag value or a positional argument.

With `PassUnknownFlagToArgs`, we do not need to classify `somevalue`; we simply leave the argument stream intact and let the caller receive everything as positional args. So we don't need to call `stripUnknownFlagValue()`.

With `IgnoreUnknownFlag`, the existing behavior treats the next token as a flag value (when it does not look like a flag) and calls `stripUnknownFlagValue()` to strip it. This behavior is inherited from the existing `IgnoreUnknownFlag` implementation.

File: flag.go (hunk 5 above)
Comment (ikedam) [reply flag.go:1155]: Same here, differs in the next call.

For an unknown flag, `parseShortArg` only need to choose between:

* strip only the first letter, or
* strip all the argument

Both `-fg=g` and `-fgh` are the former case: only strips the first letter (`f`) as an unknown flag, and pass the rest (`g=g` or `gh`).
The rest will be handled in the next `parseSingleShortArg()` call.

In the next `parseSingleShortArg()`, `g=g` results the latter case, and `gh` results the former case.
